cmd/agent/daemon/state/container_events_group.go

```diff
@@ -0,0 +1,349 @@
+	containerEventMapper    containerEventMapper
+	containerEventsSender   ContainerEventsSender
+	eventsEnrichmentService eventsEnrichmentService
+	ebpfEventNameGetter     ebpfEventNameGetter
+	currentTimeFunc         func() time.Time
+
+	batch                *castpb.ContainerEventsBatch
+	fingerprintsDigest   *xxhash.Digest
+	fingerprints         freelru.Cache[uint64, struct{}]
+	eventsQueue          chan *ebpftypes.Event
+	signatureEventsQueue chan signature.Event
+	flushedAt            time.Time
+
+	stopOnce sync.Once
+	stop     chan struct{}
+}
+
+func (g *containerEventsGroup) run(ctx context.Context) error {
+	g.log.Infof("running container events group, container=%s(%s)", g.batch.ContainerName, g.batch.ContainerId)
+	defer g.log.Infof("stopping container events group, container=%s(%s)", g.batch.ContainerName, g.batch.ContainerId)
+
+	ticker := time.NewTicker(g.cfg.flushInterval)
+	defer ticker.Stop()
+
+	runtime.SetFinalizer(g, func(s *containerEventsGroup) {
+		g.close()
+	})
+
+	for {
+		select {
+		case <-g.stop:
+			g.sendBatch(ctx, g.currentTimeFunc())
+			return nil
+		case <-ctx.Done():
+			return ctx.Err()
+		case e := <-g.eventsQueue:
+			g.handleEvent(ctx, e, nil)
+		case e := <-g.signatureEventsQueue:
+			g.handleEvent(ctx, e.EbpfEvent, e.SignatureEvent)
+		case <-ticker.C:
+			g.flush(ctx)
+		}
+	}
+}
+
+func (g *containerEventsGroup) pushEvent(event *ebpftypes.Event) {
+	select {
+	case g.eventsQueue <- event:
+	default:
+		if g.log.IsEnabled(slog.LevelDebug) {
+			g.log.Debugf("dropping event %s", g.ebpfEventNameGetter.GetEventName(event.Context.EventID))
+		}
+		metrics.AgentExporterQueueDroppedTotal.WithLabelValues("container_event").Inc()
+	}
+}
+
+func (g *containerEventsGroup) pushSignatureEvent(event signature.Event) {
+	select {
+	case g.signatureEventsQueue <- event:
+	default:
+		if g.log.IsEnabled(slog.LevelDebug) {
+			g.log.Debugf("dropping signature event %s", g.ebpfEventNameGetter.GetEventName(event.EbpfEvent.Context.EventID))
+		}
+		metrics.AgentExporterQueueDroppedTotal.WithLabelValues("container_signature_event").Inc()
+	}
+}
+
+func (g *containerEventsGroup) close() {
+	g.stopOnce.Do(func() {
+		close(g.stop)
+	})
+}
+
+func (g *containerEventsGroup) handleEvent(ctx context.Context, e *ebpftypes.Event, signatureEvent *castpb.SignatureEvent) {
+	if !g.shouldAddEvent(e) {
+		metrics.AgentSkippedEventsTotal.WithLabelValues(g.ebpfEventNameGetter.GetEventName(e.Context.EventID)).Inc()
+		return
+	}
+
+	protoContainerEvent := &castpb.ContainerEvent{}
+	// Set fields on final proto event and enrich if needed.
+	g.containerEventMapper(protoContainerEvent, e, signatureEvent)
+	g.eventsEnrichmentService.Enrich(ctx, e, protoContainerEvent)
+	g.batch.Items = append(g.batch.Items, protoContainerEvent)
+
+	now := g.currentTimeFunc()
+	if g.shouldSendBatch(now) {
+		g.sendBatch(ctx, now)
+	}
+}
+
+func (g *containerEventsGroup) shouldAddEvent(e *ebpftypes.Event) bool {
+	fingerprint, ok := g.getEventFingerprint(e)
```

Comment:
The whole fingerprinting and should drop logic is best moved to the beginning of the tracer pipeline to be honest. We need to think how we handle signatures though.

Reply:
Thanks, will create separate brunch to try this out.